### DISCUSSION TITLE
ModelSerializer uniqueness validation does not work for partial unique constraints

### DISCUSSION THREAD

**Discussion by elliot-wilson** [encode/django-rest-framework#9411] [Potential Issue]
I have a model with a relatively common partial uniqueness constraint pattern: there can be many `Person` instances with a duplicate `name` value, as long as only one of them `is_active: True`. 

However, when using a `ModelSerializer` to validate data for that model, the uniqueness validation returns `False` for _otherwise valid_ data if there is already a `Person` with that `name` in the database, even if the data has `is_active: False`.

Here's a dummy example:

```python
from django.db import models
from django.db.models import Q
from rest_framework import serializers

class Person(models.Model):
    name = models.CharField(max_length=100)
    is_active = models.BooleanField(default=False)

    class Meta:
        constraints = [
            models.UniqueConstraint(fields=["name"], condition=Q(is_active=True), name="unique_active_name"),
        ]

class PersonSerializer(serializers.ModelSerializer):
    class Meta:
        model = Person
        fields = "__all__"

Person.objects.create(name="John Doe", is_active=True)

data = {
    "name": "John Doe",
    "is_active": False,
}

Person.objects.create(**data)  # creates new user just fine

serializer = PersonSerializer(data=data)
serializer.is_valid(raise_exception=True)  # rest_framework.exceptions.ValidationError: {'name': [ErrorDetail(string='person with this name already exists.', code='unique')]}
```

The relevant source code is here:

https://github.com/encode/django-rest-framework/blob/36d5c0e74f562cbe3055f0d20818bd48d3c32359/rest_framework/utils/field_mapping.py#L65-L85

and

https://github.com/encode/django-rest-framework/blob/36d5c0e74f562cbe3055f0d20818bd48d3c32359/rest_framework/validators.py#L63-L74

Basically, the code is asking, "Is there already an `is_active: True` `Person` with this `name`?", but I believe the correct question should be more like "Is there already a record in the database that matches the constraint condition AND does the current instance match the condition?"

**Comment by elliot-wilson** [comment]
As a quick workaround, I was able to do:

```python
class PersonSerializer(serializers.ModelSerializer):
    provider = serializers.CharField(validators=[])

    class Meta:
        model = Person
        fields = "__all__"

    def validate(self, data):
        name = data.get("name")
        is_active = data.get("is_active")
        instance = self.instance

        if is_active:
            if Person.objects.filter(name=name, is_active=True).exclude(id=instance.id).exists():
                raise serializers.ValidationError({"name": "There is already an active person with this name."})

        return data
```

but it seems a little hacky to be reimplementing the constraint in the serializer when it already exists on the model.

I'd love to know what people think!

**Comment by browser-bug** [comment]
I'm experiencing the same issue. Has this been confirmed and will (potentially) get fixed or not? Thanks
